akka-projection-grpc/src/main/scala/akka/persistence/EventWriter.scala

```diff
@@ -0,0 +1,197 @@
+                }
+
+              case _ =>
+                // ignore all other journal protocol messages
+                Behaviors.same
+            }
+
+          Behaviors.receiveMessage {
+            case Write(persistenceId, sequenceNumber, event, metadata, replyTo) =>
+              val repr = PersistentRepr(
+                event,
+                persistenceId = persistenceId,
+                sequenceNr = sequenceNumber,
+                manifest = "", // adapters would be on the producing side, already applied
+                writerUuid = writerUuid,
+                sender = akka.actor.ActorRef.noSender)
+
+              val reprWithMeta = metadata match {
+                case Some(meta) => repr.withMetadata(meta)
+                case _          => repr
+              }
+
+              val newStateForPid =
+                perPidWriteState.get(persistenceId) match {
+                  case None =>
+                    if (context.log.isTraceEnabled)
+                      context.log.traceN(
+                        "Writing event persistence id [{}], sequence nr [{}], payload {}",
+                        persistenceId,
+                        sequenceNumber,
+                        event)
+                    val write = AtomicWrite(reprWithMeta) :: Nil
+                    journal ! JournalProtocol.WriteMessages(write, context.self.toClassic, context.self.path.uid)
+                    StateForPid(Map(reprWithMeta.sequenceNr -> replyTo), emptyWaitingForWrite)
+                  case Some(state) =>
+                    // write in progress for pid, add write to batch and perform once current write completes
+                    if (state.waitingForWrite.size == maxBatchSize) {
+                      replyTo ! StatusReply.error(
+                        s"Max batch reached for pid $persistenceId, at most $maxBatchSize writes for " +
+                        "the same pid may be in flight at the same time")
+                      state
+                    } else {
+                      if (context.log.isTraceEnabled)
+                        context.log.traceN(
+                          "Writing event in progress for persistence id [{}], adding sequence nr [{}], payload {} to batch",
+                          persistenceId,
+                          sequenceNumber,
+                          event)
+                      state.copy(waitingForWrite = state.waitingForWrite :+ ((reprWithMeta, replyTo)))
+                    }
+                }
+              perPidWriteState = perPidWriteState.updated(persistenceId, newStateForPid)
+              Behaviors.same
+
+            case response: JournalProtocol.Response => handleJournalResponse(response)
+
+            case unexpected =>
+              context.log.warn("Unexpected message sent to EventWriter [{}], ignored", unexpected.getClass)
+              Behaviors.same
+
+          }
+        })
+      .onFailure[Exception](SupervisorStrategy.restart)
```

Comment:
Honestly not quite sure how this would crash, other than some bug in the batching logic, errors persisting events are fed back to the sender and shouldn't crash the writer itself.

Reply:
Right, we are not failing the writer actor when we there are db errors. The CircuitBreaker in `AsyncWriteJournal` will protect the db from overload when it's already in trouble. Ok, we can leave it as plain `restart`.